packages/babel-plugin-transform-block-scoping/src/index.js

```diff
@@ -433,17 +433,47 @@ class BlockScoping {
         ]);
 
         if (violation.isAssignmentExpression()) {
-          violation
-            .get("right")
-            .replaceWith(
-              t.sequenceExpression([throwNode, violation.get("right").node]),
+          const { operator } = violation.node;
+          if (["=", "&&=", "||=", "??="].includes(operator)) {
+            violation
+              .get("right")
+              .replaceWith(
+                t.sequenceExpression([violation.get("right").node, throwNode]),
+              );
+          } else {
+            violation.replaceWith(
+              t.sequenceExpression([
+                t.binaryExpression(
+                  operator.slice(0, -1),
+                  violation.get("left").node,
+                  violation.get("right").node,
+                ),
+                throwNode,
+              ]),
             );
+          }
         } else if (violation.isUpdateExpression()) {
           violation.replaceWith(
-            t.sequenceExpression([throwNode, violation.node]),
+            t.sequenceExpression([
+              t.binaryExpression(
+                violation.node.operator[0],
+                violation.get("argument").node,
+                t.numericLiteral(1),
+              ),
```

Comment:
I think we can simplify this to `t.unaryExpression("+", violation.get("argument").node)` (even if it saves literally a single byte).

Reply:
Yes, `+c` has same effect of triggering `.valueOf()` as `c + 1`. Have added a commit to make this change.

I hadn't really bothered with optimization since in most cases code like this is a bug in any case.

Have also added another commit for the other most obvious optimization - shorten `c = f()` to just `f()` since the assignment will never be executed anyway. There are other optimizations which could be made (`a += 'long_string'` to `a + ''` for example), but in my view going further would add complexity for little gain.

Reply:
Added one more commit to shorten `c &&= x++` -> `c && x++`.